pkg/service/controller/ingressip/controller.go

```diff
@@ -0,0 +1,592 @@
+
+	ip := net.ParseIP(ipString)
+	if ip == nil {
+		// An invalid ingress ip is assigned to the service.
+		if !ic.syncProcessed {
+			// Fix the problem after processing the initial sync to
+			// avoid delaying cache updates any longer than necessary.
+			ic.pendingAllocations = append(ic.pendingAllocations, service)
+			return nil
+		} else {
+			return ic.clearPersistedAllocation(sCopy)
+		}
+	}
+
+	ipKey, ok := ic.allocationMap[ipString]
+	if ok && ipKey == key {
+		// Allocation is for this service.
+		if !ic.syncProcessed {
+			// Avoid ensuring an external ip until after the initial sync.
+			ic.pendingAllocations = append(ic.pendingAllocations, service)
+			return nil
+		} else {
+			// Ensure that the ingress ip is present in the service's spec.
+			return ic.ensureExternalIP(sCopy, ipString)
+		}
+	}
+
+	err = ic.ipAllocator.Allocate(ip)
+	if err != nil {
+		if !ic.syncProcessed {
+			// Fix the problem after processing the initial sync
+			ic.pendingAllocations = append(ic.pendingAllocations, service)
+			return nil
+		} else {
+			// Clear the allocation so the service can be re-allocated
+			// when the update is seen.
+			return ic.clearPersistedAllocation(sCopy)
+		}
+	}
+	ic.allocationMap[ipString] = key
+	if !ic.syncProcessed {
+		// Avoid ensuring an external ip until after the initial sync.
+		ic.pendingAllocations = append(ic.pendingAllocations, service)
+		return nil
+	} else {
+		// Ensure that the ingress ip is present in the service's spec.
+		return ic.ensureExternalIP(sCopy, ipString)
+	}
+}
+
+// clearPersistedAllocation ensures there is no ingress ip in the
+// service's spec and that the service's status is cleared.
+func (ic *IngressIPController) clearPersistedAllocation(service *kapi.Service) error {
+	// Assume it is safe to modify the service without worrying about changing the local cache
+
+	// An ingress ip is only allowed in ExternalIPs when a
+	// corresponding status exists, so update the spec first to avoid
+	// failing admission control.
+	ingressIP := service.Status.LoadBalancer.Ingress[0].IP
+
+	filteredIPs := service.Spec.ExternalIPs[:0]
+	changed := false
+	for _, ip := range service.Spec.ExternalIPs {
+		if ip != ingressIP {
+			filteredIPs = append(filteredIPs, ip)
+		} else {
+			changed = true
+		}
+	}
+	if changed {
+		service.Spec.ExternalIPs = filteredIPs
+		err := ic.persistSpec(service)
+		if err != nil {
+			return err
+		}
+	}
+
+	service.Status.LoadBalancer = kapi.LoadBalancerStatus{}
+	return ic.persistStatus(service)
+}
+
+// allocate assigns an unallocated ip to the given service and updates
+// the service's persisted state.
+func (ic *IngressIPController) allocate(service *kapi.Service, key string) error {
+	if !ic.syncProcessed {
+		ic.pendingAllocations = append(ic.pendingAllocations, service)
+		return nil
+	}
+
+	// Make a copy to avoid mutating cache state
+	t, err := kapi.Scheme.DeepCopy(service)
+	if err != nil {
+		return err
+	}
+	sCopy := t.(*kapi.Service)
```

Comment:
style, just set this to `service = t.(*kapi.Service)` to prevent later code from accidentally changing status.  `sCopy` is unusual within our style guides (no app-hungarian here)